Concurrent update embedded caches and database

pruivo opened the pull request:
Closes #42374

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

- [x] Javadoc
- [x] Replace `SerializeExecutionsByKey` with `ActionSequencer` (non-blocking)
- [x] Consider dropping `SessionChangesPerformer` as it has 1 implementation only
- [x] Code cleanup

pruivo committed:
Concurrent update embedded caches and database

Closes #42374

Signed-off-by: Pedro Ruivo <1492066+pruivo@users.noreply.github.com>

pruivo committed:
Remove SessionChangesPerformer

Signed-off-by: Pedro Ruivo <1492066+pruivo@users.noreply.github.com>

pruivo committed:
action sequencer

Signed-off-by: Pedro Ruivo <1492066+pruivo@users.noreply.github.com>

pruivo committed:
code cleanup + javadoc

Signed-off-by: Pedro Ruivo <1492066+pruivo@users.noreply.github.com>

pruivo committed:
fix tracing

Signed-off-by: Pedro Ruivo <1492066+pruivo@users.noreply.github.com>

pruivo committed:
rename CacheInfo -> CacheHolder

Signed-off-by: Pedro Ruivo <1492066+pruivo@users.noreply.github.com>

pruivo committed:
review

Signed-off-by: Pedro Ruivo <1492066+pruivo@users.noreply.github.com>

pruivo commented:
@ryanemerson updated!

ryanemerson commented on model/infinispan/src/main/java/org/keycloak/infinispan/module/factory/OpenTelemetrySpan.java:
```diff
@@ -11,14 +11,13 @@
         // This should be fixed in ISPN 16+ so that it is no longer needed
         // https://github.com/infinispan/infinispan/issues/15287
-        this.scope = span.makeCurrent();
+        //this.scope = span.makeCurrent();
```
Thanks for explaining. I think we're good then as this isn't a regression, given that we introduced JGroups tracing in 26.4 #39658.

ahus1 committed:
Loading a single session should now run a delete

Running it on the whole table is slow. Running it via the EM will cause concurrency problems. Running a batch delete might cause stale objects.

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 reviewed (approved):
Thank you for this extensive change and all the reasoning you and everyone else did around this issue!
